lib/elixir/lib/module/types/descr.ex

```diff
@@ -728,27 +728,51 @@ defmodule Module.Types.Descr do
   end
 
   # Intersects two map literals; throws if their intersection is empty.
-  defp map_literal_intersection(tag1, map1, tag2, map2) do
-    default1 = map_tag_to_type(tag1)
-    default2 = map_tag_to_type(tag2)
+  # Both open: the result is open.
+  defp map_literal_intersection(:open, map1, :open, map2) do
+    new_fields =
+      Map.merge(map1, map2, fn _, type1, type2 -> non_empty_intersection!(type1, type2) end)
```

Comment:
`:maps.merge_with` and `:maps.intersect_with` care about the order in the combiner function, which we don't here. That's why we created `union_static`  and `intersection_static`, shall we use these functions instead?

Reply:
In the current version which does not have a different fallback type it should be fine!
Will try to do some benchmark as well to have an idea.

Reply:
OK, [updated](https://github.com/sabiwara/elixir_benches/commit/4f8724114c1801437a904e1f6f0c7c01c3a62d54) the bench to try the "static" merge approach. In most cases it doesn't seem to make a noticeable difference.

I also noticed that we cannot reuse these functions as is due to subtle differences:
- `union_static` is a "merge" as we want, but performs a union on types, not an intersection as we want
- we also need to throw if the intersection is empty

We'd need more specialized function which I'm not sure is worth it, or we could make the *_static functions less specialized by accepting a callback.

Reply:
@sabiwara I am ok with making them less generic. we introduce our own merge and intersection functions :+1:

Reply:
Sounds good. How should we name them?
`symmetrical_merge(left, right, fun)` & `symmetrical_intersection(left, right, fun)`? (since it's about not caring about the order of parameters)

Reply:
perfect!

Reply:
Done in [731f542](https://github.com/elixir-lang/elixir/pull/13524/commits/731f5420cce87ed36984599310b3662285bdbea7)